// @hono/auth-js/dist/index.cjs
const require_chunk = require('./chunk-CUT6urMc.cjs');
let __auth_core = require("@auth/core");
__auth_core = require_chunk.__toESM(__auth_core);
let hono_adapter = require("hono/adapter");
hono_adapter = require_chunk.__toESM(hono_adapter);
let hono_http_exception = require("hono/http-exception");
hono_http_exception = require_chunk.__toESM(hono_http_exception);

//#region src/index.ts
function setEnvDefaults(env$1, config) {
	config.secret ??= env$1.AUTH_SECRET;
	(0, __auth_core.setEnvDefaults)(env$1, config);
}
function reqWithEnvUrl(req, authUrl) {
	if (authUrl) {
		const reqUrlObj = new URL(req.url);
		const authUrlObj = new URL(authUrl);
		for (const prop of [
			"hostname",
			"protocol",
			"port",
			"password",
			"username"
		]) if (authUrlObj[prop]) reqUrlObj[prop] = authUrlObj[prop];
		return new Request(reqUrlObj.href, req);
	}
	const url = new URL(req.url);
	const newReq = new Request(url.href, req);
	const proto = newReq.headers.get("x-forwarded-proto");
	const host = newReq.headers.get("x-forwarded-host") ?? newReq.headers.get("host");
	if (proto != null) url.protocol = proto.endsWith(":") ? proto : `${proto}:`;
	if (host != null) {
		url.host = host;
		const portMatch = host.match(/:(\d+)$/);
		if (portMatch) url.port = portMatch[1];
		else url.port = "";
		newReq.headers.delete("x-forwarded-host");
		newReq.headers.delete("Host");
		newReq.headers.set("Host", host);
	}
	return new Request(url.href, newReq);
}
async function getAuthUser(c) {
	const config = c.get("authConfig");
	const ctxEnv = (0, hono_adapter.env)(c);
	setEnvDefaults(ctxEnv, config);
	const authReq = reqWithEnvUrl(c.req.raw, ctxEnv.AUTH_URL);
	const origin = new URL(authReq.url).origin;
	const request = new Request(`${origin}${config.basePath}/session`, { headers: { cookie: c.req.header("cookie") ?? "" } });
	let authUser = {};
	return (await (await (0, __auth_core.Auth)(request, {
		...config,
		callbacks: {
			...config.callbacks,
			async session(...args) {
				authUser = args[0];
				const session = await config.callbacks?.session?.(...args) ?? args[0].session;
				return {
					user: args[0].user ?? args[0].token,
					...session
				};
			}
		}
	})).json())?.user ? authUser : null;
}
function verifyAuth() {
	return async (c, next) => {
		const authUser = await getAuthUser(c);
		if (!(!!authUser?.token || !!authUser?.user)) throw new hono_http_exception.HTTPException(401, { res: new Response("Unauthorized", { status: 401 }) });
		c.set("authUser", authUser);
		await next();
	};
}
function initAuthConfig(cb) {
	return async (c, next) => {
		const config = await cb(c);
		c.set("authConfig", config);
		await next();
	};
}
function authHandler() {
	return async (c) => {
		const config = c.get("authConfig");
		const ctxEnv = (0, hono_adapter.env)(c);
		setEnvDefaults(ctxEnv, config);
		if (!config.secret || config.secret.length === 0) throw new hono_http_exception.HTTPException(500, { message: "Missing AUTH_SECRET" });
		const body = c.req.raw.body ? await c.req.blob() : void 0;
		const res = await (0, __auth_core.Auth)(reqWithEnvUrl(new Request(c.req.raw.url, {
			body,
			cache: c.req.raw.cache,
			credentials: c.req.raw.credentials,
			headers: c.req.raw.headers,
			integrity: c.req.raw.integrity,
			keepalive: c.req.raw.keepalive,
			method: c.req.raw.method,
			mode: c.req.raw.mode,
			redirect: c.req.raw.redirect,
			referrer: c.req.raw.referrer,
			referrerPolicy: c.req.raw.referrerPolicy,
			signal: c.req.raw.signal
		}), ctxEnv.AUTH_URL), config);
		return new Response(res.body, res);
	};
}

//#endregion
exports.authHandler = authHandler;
exports.getAuthUser = getAuthUser;
exports.initAuthConfig = initAuthConfig;
exports.reqWithEnvUrl = reqWithEnvUrl;
exports.setEnvDefaults = setEnvDefaults;
exports.verifyAuth = verifyAuth;
